### DISCUSSION TITLE
static "extra-lazy" operations for collections

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#3779] [Design Proposals]
~I would like to remove extra-lazy collections from Hibernate 6.~

~But instead of just killing them without providing any way to achieve the same thing,~ we could easily add some operations like:

```
Hibernate.getSize(my.getItems())
Hibernate.get(my.getMap(), key)
Hibernate.contains(my.getItems(), item)
Hibernate.remove(my.getItems(), item)
Hibernate.put(my.getMap(), key, value)
```

So that you can do item-level operations on a collection without actually loading it.

Reactions to this idea?

FTR I don't think it would be hard for me to implement it. OTOH, it's code, and code must be maintained.

**Comment by Sanne** [comment]
I do like the new helpers you're proposing, but I'm skeptical of removing the extra-lazy capability.

So reasons to keep the extra-lazy:

- People know it, and it's a useful feature even in its current shape (albeit limited)
- When mapping a relation as a `Map<k,v>` it allows loading selected keys only, even by code which is oblivious of the object being managed, and not having to depend on Hibernate.
- It allows loading patterns which are beneficial to make optimal use of the 2LC

Reasons to introduce such a new API as well (eventually?):

- It allows more fine grained control, e.g. get only the _size_ without having to load it all, and we could implement `contains` as a boolean-returning SQL w/o actually loading the entry (not sure if that's actually a good idea, my point being we'd have more such options)
- It allows performing such optimisations even if the extra-lazyness wasn't specificed on the mapping; it's great to focus more on the use case rather than enforce a single option via static mapping.

**Reply by beikov** [reply]
So this feature is here to stay as we figured it is indeed useful and people actually use it.

**Reply by gavinking** [reply]
Right, it seems so. But the proposal to add operations to the `Hibernate` class is still live.

**Comment by gavinking** [comment]
This is implemented in https://github.com/hibernate/hibernate-orm/pull/5094.
